File: core/rule.py
from collections import Counter
import logging
import json

logger = logging.getLogger('ddz')

'''
# A 2 3 4 5 6 7 8 9 0 J Q K W w
# 记分
# 地主先出完所有的牌就赢了，如果没有出过炸弹或火箭，那么每个农民要把定约的分数(1分、2分或3分)付给地主。
# 两个农民中有一个先出完所有的牌，地主就输了，那么地主要把定约的分数付给每个农民。
# 每当任何一个玩家出了炸弹或火箭，那么分数就要翻一番。
# 例如某局牌出了2个炸弹和1个火箭，叫3分的地主如果先出完，他就向每个农民赢得24分【总共赢得48分】，
# 如果农民先出完，地主就向每个农民输掉24分【总共输掉48分】。
'''

CARD_TYPES = [
    # 'rocket', 'bomb',
    'single', 'pair', 'trio', 'trio_pair', 'trio_single',
    'seq_single5', 'seq_single6', 'seq_single7', 'seq_single8', 'seq_single9', 'seq_single10', 'seq_single11',
    'seq_single12',
    'seq_pair3', 'seq_pair4', 'seq_pair5', 'seq_pair6', 'seq_pair7', 'seq_pair8', 'seq_pair9', 'seq_pair10',
    'seq_trio2', 'seq_trio3', 'seq_trio4', 'seq_trio5', 'seq_trio6',
    'seq_trio_pair2', 'seq_trio_pair3', 'seq_trio_pair4', 'seq_trio_pair5',
    'seq_trio_single2', 'seq_trio_single3', 'seq_trio_single4', 'seq_trio_single5',
    'bomb_pair', 'bomb_single'
]

with open('static/rule.json', 'r') as f:
    rules = json.load(f)


def is_contains(parent, child):
    parent, child = Counter(parent), Counter(child)
    for k, n in child.items():
        if k not in parent or n > parent[k]:
            return False
    return True


def cards_above(hand_pokers, turn_pokers):
    hand_cards = _to_cards(hand_pokers)
    turn_cards = _to_cards(turn_pokers)

    card_type, card_value = _cards_value(turn_cards)
    if not card_type:
        return []

    one_rule = rules[card_type]
    for i, t in enumerate(one_rule):
        if i > card_value and is_contains(hand_cards, t):
            return _to_pokers(hand_pokers, t)

    if card_value < 1000:
        one_rule = rules['bomb']
        for t in one_rule:
            if is_contains(hand_cards, t):
                return _to_pokers(hand_pokers, t)
        if is_contains(hand_cards, 'wW'):
            return [52, 53]
    return []


def _to_cards(pokers):
    cards = []
    for p in pokers:
        if p == 52:
            cards.append('W')
        elif p == 53:
            cards.append('w')
        else:
            cards.append('A234567890JQK'[p % 13])
    return _sort_card(cards)


def _to_poker(card):
    if card == 'W':
        return [52]
    if card == 'w':
        return [53]

    cards = 'A234567890JQK'
    for i, c in enumerate(cards):
        if c == card:
            return [i, i + 13, i + 13*2, i + 13*3]
    return [54]


def _to_pokers(hand_pokers, cards):
    pokers = []
    for card in cards:
        candidates = _to_poker(card)
        for cd in candidates:
            if cd in hand_pokers and cd not in pokers:
                pokers.append(cd)
                break
    return pokers


def _cards_value(cards):
    cards = ''.join(cards)
    if cards == 'wW':
        return 'rocket', 2000

    value = _index_of(rules['bomb'], cards)
    if value >= 0:
        return 'bomb', 1000 + value

    return _card_type(cards)


def compare_poker(a_pokers, b_pokers):
    if not a_pokers or not b_pokers:
        if a_pokers == b_pokers:
            return 0
        if a_pokers:
            return 1
        if b_pokers:
            return 1

    a_card_type, a_card_value = _cards_value(_to_cards(a_pokers))
    b_card_type, b_card_value = _cards_value(_to_cards(b_pokers))
    if a_card_type == b_card_type:
        return a_card_value - b_card_value

    if a_card_value >= 1000:
        return 1
    else:
        return 0


def _sort_card(cards):
    cards.sort(key=lambda ch: '34567890JQKA2wW'.index(ch))
    return cards


def _index_of(array, ele):
    if len(array[0]) != len(ele):
        return -1
    for i, e in enumerate(array):
        if e == ele:
            return i
    return -1


def _card_type(cards):
    for t in CARD_TYPES:
        value = _index_of(rules[t], cards)
        if value >= 0:
            return t, value
    logger.info('Unknown Card Type: %s', cards)
    # raise Exception('Unknown Card Type: %s' % cards)
    return '', 0

File: doudizhu/static/js/rule.js
PG.Poker = function (game, id, frame) {

    Phaser.Sprite.call(this, game, game.world.width / 2, game.world.height * 0.4, 'poker', frame - 1);

    this.anchor.set(0.5);
    this.id = id;

    return this;
};

PG.Poker.prototype = Object.create(Phaser.Sprite.prototype);
PG.Poker.prototype.constructor = PG.Poker;

PG.Poker.comparePoker = function (a, b) {

    if (a instanceof Array) {
        a = a[0];
        b = b[0];
    }


    if (a > 52 || b > 52) {
        return -(a - b);
    }
    a = a % 13;
    b = b % 13;
    if (a <= 2) {
        a += 13;
    }
    if (b <= 2) {
        b += 13;
    }
    return -(a - b);
};

PG.Poker.toCards = function (pokers) {
    let cards = [];
    for (let i = 0; i < pokers.length; i++) {
        let pid = pokers[i];
        if (pid instanceof Array) {
            pid = pid[0];
        }
        if (pid === 53) {
            cards.push('w');
        } else if (pid === 54) {
            cards.push('W');
        } else {
            cards.push("KA234567890JQ"[pid % 13]);
        }
    }
    return cards;

};

PG.Poker.canCompare = function (pokersA, pokersB) {
    let cardsA = this.toCards(pokersA);
    let cardsB = this.toCards(pokersB);
    return PG.Rule.cardsValue(cardsA)[0] == PG.Rule.cardsValue(cardsB)[0];
};

PG.Poker.toPokers = function (pokerInHands, cards) {
    let pokers = [];
    for (let i = 0; i < cards.length; i++) {
        let candidates = this.toPoker(cards[i]);
        for (let j = 0; j < candidates.length; j++) {
            if (pokerInHands.indexOf(candidates[j]) != -1 && pokers.indexOf(candidates[j]) == -1) {
                pokers.push(candidates[j]);
                break
            }
        }
    }
    return pokers;
};

PG.Poker.toPoker = function (card) {

    const cards = "?A234567890JQK";
    for (let i = 1; i < cards.length; i++) {
        if (card === cards[i]) {
            return [i, i + 13, i + 26, i + 39];
        }
    }
    if (card === 'w') {
        return [53];
    } else if (card === 'W') {
        return [54];
    }
    return [55];

};

PG.Rule = {};

PG.Rule.cardsAbove = function (handCards, turnCards) {

    let turnValue = this.cardsValue(turnCards);
    if (turnValue[0] === '') {
        return '';
    }
    handCards.sort(this.sorter);
    let oneRule = PG.RuleList[turnValue[0]];
    for (let i = turnValue[1] + 1; i < oneRule.length; i++) {
        if (this.containsAll(handCards, oneRule[i])) {
            return oneRule[i];
        }
    }

    if (turnValue[1] < 10000) {
        oneRule = PG.RuleList['bomb'];
        for (let i = 0; i < oneRule.length; i++) {
            if (this.containsAll(handCards, oneRule[i])) {
                return oneRule[i];
            }
        }
        if (this.containsAll(handCards, 'wW')) {
            return 'wW';
        }
    }

    return '';
};

PG.Rule.bestShot = function (handCards) {

    handCards.sort(this.sorter);
    let shot = '';
    for (let i = 2; i < this._CardsType.length; i++) {
        let oneRule = PG.RuleList[this._CardsType[i]];
        for (let j = 0; j < oneRule.length; j++) {
            if (oneRule[j].length > shot.length && this.containsAll(handCards, oneRule[j])) {
                shot = oneRule[j];
            }
        }
    }

    if (shot === '') {
        oneRule = PG.RuleList['bomb'];
        for (let i = 0; i < oneRule.length; i++) {
            if (this.containsAll(handCards, oneRule[i])) {
                return oneRule[i];
            }
        }
        if (this.containsAll(handCards, 'wW'))
            return 'wW';
    }

    return shot;
};

PG.Rule._CardsType = [
    'rocket', 'bomb',
    'single', 'pair', 'trio', 'trio_pair', 'trio_single',
    'seq_single5', 'seq_single6', 'seq_single7', 'seq_single8', 'seq_single9', 'seq_single10', 'seq_single11', 'seq_single12',
    'seq_pair3', 'seq_pair4', 'seq_pair5', 'seq_pair6', 'seq_pair7', 'seq_pair8', 'seq_pair9', 'seq_pair10',
    'seq_trio2', 'seq_trio3', 'seq_trio4', 'seq_trio5', 'seq_trio6',
    'seq_trio_pair2', 'seq_trio_pair3', 'seq_trio_pair4',
    'seq_trio_single2', 'seq_trio_single3', 'seq_trio_single4', 'seq_trio_single5',
    'bomb_pair', 'bomb_single'];

PG.Rule.sorter = function (a, b) {
    let card_str = '34567890JQKA2wW';
    return card_str.indexOf(a) - card_str.indexOf(b);
};

PG.Rule.index_of = function (array, ele) {
    if (array[0].length !== ele.length) {
        return -1;
    }
    for (let i = 0, l = array.length; i < l; i++) {
        if (array[i] === ele) {
            return i;
        }
    }
    return -1;
};

PG.Rule.containsAll = function (parent, child) {
    let index = 0;
    for (let i = 0; i < child.length; i++) {
        index = parent.indexOf(child[i], index);
        if (index === -1) {
            return false;
        }
        index += 1;
    }
    return true;
};

PG.Rule.cardsValue = function (cards) {

    if (typeof(cards) != 'string') {
        cards.sort(this.sorter);
        cards = cards.join('');
    }

    if (cards === 'wW')
        return ['rocket', 20000];

    let index = this.index_of(PG.RuleList['bomb'], cards);
    if (index >= 0)
        return ['bomb', 10000 + index];

    let length = this._CardsType.length;
    for (let i = 2; i < length; i++) {
        let typeName = this._CardsType[i];
        let index = this.index_of(PG.RuleList[typeName], cards);
        if (index >= 0)
            return [typeName, index];
    }
    console.log('Error: UNKNOWN TYPE ', cards);
    return ['', 0];
};

PG.Rule.compare = function (cardsA, cardsB) {

    if (cardsA.length === 0 && cardsB.length === 0) {
        return 0;
    }
    if (cardsA.length === 0) {
        return -1;
    }
    if (cardsB.length === 0) {
        return 1;
    }

    let valueA = this.cardsValue(cardsA);
    let valueB = this.cardsValue(cardsB);

    if ((valueA[1] < 10000 && valueB[1] < 10000) && (valueA[0] !== valueB[0])) {
        console.log('Error: Compare ', cardsA, cardsB);
    }

    return valueA[1] - valueB[1];
};

PG.Rule.shufflePoker = function () {
    let pokers = [];
    for (let i = 0; i < 54; i++) {
        pokers.push(i);
    }

    let currentIndex = pokers.length, temporaryValue, randomIndex;
    while (0 !== currentIndex) {
        randomIndex = Math.floor(Math.random() * currentIndex);
        currentIndex -= 1;

        temporaryValue = pokers[currentIndex];
        pokers[currentIndex] = pokers[randomIndex];
        pokers[randomIndex] = temporaryValue;
    }
    return pokers;
}
;

